tools/coqdoc/output.ml

```diff
@@ -107,6 +107,50 @@ let set_module m sub = current_module := (m,sub);
 let item_level = ref 0
 let in_doc = ref false
 
+let (%) f g x = f (g x)
+
+let string_flatmapi f =
+  let b = Buffer.create 16 in
+  fun s ->
+    Buffer.clear b;
+    String.iteri (fun i c -> Buffer.add_string b (f i c)) s;
+    Buffer.contents b
+
+let string_flatmap f = string_flatmapi (fun _ -> f)
+
+let html_escape = string_flatmap (function
+  | '<' -> "&lt;"
+  | '>' -> "&gt;"
+  | '&' -> "&amp;"
+  | '\"' -> "&quot;"
+  | c -> String.make 1 c
+  )
+
+let uri_escape = string_flatmap (function
+  | '>' | '<' | '#' | '&' as c -> sprintf "%%%X" (int_of_char c)
+  | ' ' -> "_"
+  | c -> String.make 1 c
+  )
+
+(* A 'fragment' is the part in the uri path that comes after the '#'. *)
+
+let allowed_fragment_char = function
+  | 'a'..'z' | 'A'..'Z' | '0'..'9' | '.' | '_' | '<' | '>' | '&'
+    | '\'' | '"' | '-' | ':' -> true
+  | _ -> false
+
+let uri_fragment_escape s =
+  let string_contains f s =
+     let rec find i = i < String.length s && (f s.[i] || find (i+1))
+     in find 0 in
+
+  if string_contains (not % allowed_fragment_char) s
```

Comment:
A bit too many indirections for my taste, I had a hard time reading it. Here is a suggestion:
```ocaml
let uri_fragment_escape s =
  let l = String.length s in
  let rec valid i = if i = l then true else allowed_fragment_char s.[i] && valid (i + 1) in
  if valid 0 then
```